[Update Test] Update TC IDM 4 3 python3 test module from Raul

j-ororke opened the pull request:
####  Dependencies
This PR is based on Raul's IDM-4-3 python3 test module PR #[33477](https://github.com/project-chip/connectedhomeip/pull/33477)

#### Summary

- Updating IDM_4_3 python3 test module to resolve merge conflicts in PR #[33477](https://github.com/project-chip/connectedhomeip/pull/33477/files).
- Made following changes or updates to the test code to complete this request:
1. Added XML spec-based dynamic writable attribute discovery for steps 8, 10, and 11
2. Implemented change report verification to confirm DUT sends reports for modified attributes
3. Fixed INVALID_ACTION errors by excluding problematic clusters from wildcard subscriptions as done in ACE tests.
4. Added MRP parameter negotiation for accurate timeout calculations in step 4 using information from spec and ReliableMessageProtocolConfig.h
5. Fixed AttributePath formatting for wildcard patterns (steps 8-12)
6. Integrated AttributeSubscriptionHandler and EventSubscriptionHandler patterns from event_attribute_reporting.py for robust subscription management

Test Plan Changes made in Test Plan PR #[5667](https://github.com/CHIP-Specifications/chip-test-plans/pull/5667) to help with reviewing the code changes made:
1. Removed steps  4-10 (data type subscription tests) as these are covered by TC-ACE-2.4 in PR [41273](https://github.com/project-chip/connectedhomeip/pull/41273) and simiar reasoning as mentioned here [266](https://github.com/project-chip/matter-test-scripts/issues/266#issuecomment-2962672334)
2. Moved test step 3 to test step 1
3. Changed the wording for test step 2 for basic attribute change timing validation
4. New test step 3 is original test step 11
5. Step 4 now combines test steps 12 & 13 (MinInterval/MaxInterval with empty reports)
6. Test steps 5-12 were originally test step 14-21

#### Testing

- Tested in WSL Linux (Merged in latest changes from upstream master branch) to resolve merge conflicts.
- Ran following command to validate this test passes:
```
./scripts/tests/run_python_test.py --factory-reset --app out/linux-x64-all-clusters/chip-all-clusters-app --app-args "--discriminator 1234 --KVS kvs1" --script-args "--storage-path admin_storage.json --commissioning-method on-network --discriminator 1234 --passcode 20202021 --PICS src/app/tests/suites/certification/ci-pics-values --endpoint 0" --script src/python_testing/TC_IDM_4_3.py
```

#### Related Issues
Raul's PR with initial implementation: [33477](https://github.com/project-chip/connectedhomeip/pull/33477)
- This added the very helpful SetNotifySubscriptionStillActiveCallback() for empty subscription report mechanism in C++ and plumbing in the python framework with timing validation that is utilized in this test module.

Original Test plan link: https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/interactiondatamodel.adoc#443-tc-idm-43-report-data-messages-post-subscription-activation-from-dut-test-cases-dut_server

New Test Plan PR with changes made: [5667](https://github.com/CHIP-Specifications/chip-test-plans/pull/5667)

#### Notes
This PR finishes the test module code upon foundational work done by Raul in the initial stages of this test in PR 33477, who created the SetNotifySubscriptionStillActiveCallback() mechanism in both the C++ and Python framework layers to enable empty report detection and timing validation in subscription tests.

raul-marquez-csa committed:
Initial commit

raul-marquez-csa committed:
progress

raul-marquez-csa committed:
Merge branch 'project-chip:master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Updates steps to step array

raul-marquez-csa committed:
Fix restyle

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
progress

raul-marquez-csa committed:
Step 2 progress

raul-marquez-csa committed:
Lint

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Fix restyle

raul-marquez-csa committed:
Merge branch 'tc-idm-4.3-(python)' of github.com:raul-marquez-csa/connectedhomeip into tc-idm-4.3-(python)

raul-marquez-csa committed:
Merge branch 'project-chip:master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Adds handleNotifySubscriptionStillActive to AsyncReadTransaction, temporary debug logging

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Adds SetNotifySubscriptionStillActiveCallback

raul-marquez-csa committed:
Merge branch 'tc-idm-4.3-(python)' of github.com:raul-marquez-csa/connectedhomeip into tc-idm-4.3-(python)

raul-marquez-csa committed:
Measuring time

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
progress

raul-marquez-csa committed:
Merge branch 'tc-idm-4.3-(python)' of github.com:raul-marquez-csa/connectedhomeip into tc-idm-4.3-(python)

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Merge branch 'project-chip:master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Removes duplicate AttributeChangeCallback class

raul-marquez-csa committed:
progress

raul-marquez-csa committed:
Merge branch 'project-chip:master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Step 1

raul-marquez-csa committed:
Merge branch 'project-chip:master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Merge branch 'project-chip:master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Step2 progress

raul-marquez-csa committed:
Step2 progress

raul-marquez-csa committed:
Merge remote-tracking branch 'origin/master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

raul-marquez-csa committed:
Update to latest master merge and refactors

raul-marquez-csa committed:
Fix restyle

raul-marquez-csa committed:
Fix restyle

raul-marquez-csa committed:
Merge branch 'master' into tc-idm-4.3-(python)

j-ororke committed:
Merge remote-tracking branch 'upstream/master' into update_TC_IDM_4_3_test

j-ororke committed:
Updating imports in TC_IDM_4_3 python3 test module

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Resolving linting error

j-ororke committed:
Resolving linting errors and removing unused imports

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
TC_IDM_4_3: Implement and validate subscription test steps 3-17:
- Implemented all remaining test steps (3-17) with proper validation
- Added XML spec-based dynamic writable attribute discovery for steps 13, 15, and 16
- Implemented change report verification to confirm DUT sends reports for modified attributes
- Fixed INVALID_ACTION errors by excluding problematic clusters (AccessControl, NetworkCommissioning,
  CameraAvStreamManagement, OperationalCredentials, ValveConfigurationAndControl,
  CarbonDioxideConcentrationMeasurement) from wildcard subscriptions
- Added MRP parameter negotiation for accurate timeout calculations in step 9 using information from spec and ReliableMessageProtocolConfig.h
- Fixed subscription callback initialization to prevent AttributeError
- Fixed AttributePath formatting for wildcard patterns (steps 14-17)
- Added list attribute detection using typing.get_origin() for typed lists
- Integrated AttributeSubscriptionHandler and EventSubscriptionHandler patterns from
  event_attribute_reporting.py for robust subscription management

j-ororke committed:
Restyled by autopep8

j-ororke committed:
Resolving linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
TC-IDM-4.3: Remove data type tests and realign test steps with implementation

- Remove steps 2-7: Data type subscription tests covered by TC-ACE-2.2
- Readd original step 2 with different wording: Single attribute change with MinInterval/MaxInterval timing validation
- Renumber steps 3-12: Align with actual implementation
- Update descriptions: Match test flow and add original test plan step mappings to test steps so they are easier to review
- Step 4 combines original steps 12+13: Empty reports and attribute change timing

j-ororke committed:
Minor update to comment wording

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving linting errors

j-ororke committed:
removed unused function

j-ororke committed:
Resolving linting errors

j-ororke committed:
Added some more clusters to skip in test steps 10 and 11 to be safe, also added better logic handling to assert if attribute values returned do not match our expected values during wildcard testing

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving linting error

j-ororke committed:
Removing some unneccessary additional sleep time

cecille commented on src/controller/python/matter/clusters/Attribute.py:
```diff
@@ -867,9 +876,20 @@ def handleReportBegin(self):
-        # self._event_loop.call_soon_threadsafe(self._handleReportEnd)
         self._handleReportEnd()
 
+    # def _handleNotifySubscriptionStillActive(self):
```
why are these commented out?

j-ororke replied:
Sorry, I should have noticed this sooner. Thanks for catching this!
These are leftover dead code from when Raul was building out the SetNotifySubscriptionStillActiveCallback() infrastructure. Really appreciate all the foundational work he did to get this working, it made implementing the remainder of TC-IDM-4.3 much easier.
I just cleaned up those commented lines here now. Thank you for the review!

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1097 @@
+Full test plan link for details:
+https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/interactiondatamodel.adoc#443-tc-idm-43-report-data-messages-post-subscription-activation-from-dut-test-cases-dut_server
+
+Note:
```
While I appreciate the effort to recognize Raul's work here, this comment is unlikely to make sense to later readers, so should probably be removed. This would be good in the PR description so it ends up in the history.

j-ororke replied:
Understood. I’ve removed this docstring here and moved that note to the PR description instead so it’s captured in the history.

cecille commented on src/controller/python/matter/clusters/Attribute.py:
```diff
@@ -584,6 +584,15 @@ def SetErrorCallback(self, callback: Callable[[int, SubscriptionTransaction], No
+        is expected to have the following signature:
+            def Callback()
+        '''
+        if callback is not None:
```
If the callback is none, do you want to unset it?

j-ororke replied:
I believe you are correct, we should support unsetting the callback when None is passed. I've updated the implementation to remove the None check and added documentation that explicitly states passing None will unset the callback. The handler already checks for truthiness before invoking, so this should work.

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1097 @@
+            callback: Optional callback function. If None, uses a no-op lambda.
+        """
+        if callback is None:
+            def callback(): return None
```
See above - suggest allowing None callback to unset the callback.

j-ororke replied:
Please see above, the function this was in is now removed and we are now initializing _notify_subscription_still_active_callback = None in the AsyncReadTransaction.__init__ method, this makes it a bit cleaner and you had suggested above.

j-ororke commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1097 @@
+        return writable_attrs
+
+    def _init_subscription_callback(self, subscription, callback=None):
+        """Initialize the subscription callback to prevent AttributeError.
```
You appear to be right that not initializing this could cause AttributeErrors in other tests. 
To avoid this happening, I've added _notify_subscription_still_active_callback = None to the AsyncReadTransaction.__init__ method so it's always initialized that way. 
This means tests that don't care about empty reports work fine without any changes, while tests like TC_IDM_4_3 can still set the callback when needed.

j-ororke commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1097 @@
+                                    'new_value': current_val
+                                })
+
+                        elif resp[0].Status == Status.UnsupportedWrite:
```
Good point, I have removed this check here now, thank you!

j-ororke committed:
Fix subscription callbacks, removed some unneeded checks for wildcard subscription writes, :

- Initialized _notify_subscription_still_active_callback to fix AttributeError
- Removed prior uncommented code in Attribute.py
- Removed prior UnsupportedWrite check in wildcard subscription write to attribute
- Removed prior init_subscription_callback() as no longer needed
- Replaced the hardcoded 0xFC00 <= cluster_id <= 0xFFFE checks with is_standard_cluster_id() in test steps 10 and 11

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving linting errors

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Updating to using nodeId var name

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Fix TC_IDM_4_3 wildcard subscription verification

- Added WildcardAttributeSubscriptionHandler to event_attribute_reporting module for queue-based report tracking
- Removed read-before-write to prevent subscription timeout
- Fixed list attribute handling (skip empty lists)
- Simplified wait logic with early exit when all reports received

j-ororke committed:
Resolving linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving minor linting errors and removing some debug logging statements

j-ororke committed:
Fix wildcard subscription verification and timeout issues:

- Set autoResubscribe=False to prevent callback loss during resubscription
- Increase Step 10 max_interval to 120s to prevent timeout during writes
- Add 1s delays after subscription shutdown to avoid race conditions
- Fix attribute value generation to ensure actual changes for all types
- Remove read-before-write to prevent subscription liveness timeout
- Add priority attributes to an exclusion list (ACL, InterfaceEnabled)
- Limit MAX_CLUSTER_PATHS_PER_SUBSCRIPTION to 25 based on device memory
  constraints during priming report generation
- Update max_changes to attributes to 20 to stay well under subscription timeout window of 120s

j-ororke committed:
Resolving linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1097 @@
+        # (This was originally test step 19 in the test plan)
+        self.step(10)
+        # Subscribe to ALL attributes from ALL clusters on ALL endpoints
+        # Note: Some clusters have attributes that don't support subscriptions (INVALID_ACTION) or only support subscriptions in special circumstances:
```
Hi Cecille,
I have investigated each cluster in the `CLUSTERS_WITH_SUBSCRIPTION_ISSUES` list here to verify whether they actually have subscription problems or if it was a test design issue.

## Investigation

1. Subscription establishment success
2. Individual attribute write operations to verify 

## Findings

After testing, **all clusters actually support subscriptions correctly**. The `INVALID_ACTION` errors have been resolved with new test logic implemented after further investigation and knowledge based understanding was commenced, this resulted in removing the prior exclusions and exclusion list below:

### Resolved:

**Switch, PowerSource, BasicInformation, Identify, All Concentration Measurement Clusters**
- **Root cause**: Test's attribute value logic had caused write failures
  - Integer attributes: `cached_val + 1` exceeded valid ranges (e.g., percentages > 100) → `ConstraintError` (Status 135)
  - String attributes: Generated strings too long for constraints (e.g., `Location` max 2 chars per spec 11.1.3.7) `ConstraintError` (Status 135)
- **Fix**: Improved value generation to respect attribute type constraints and wrap around valid ranges

**ValveConfigurationAndControl**
- **Root cause**: `DefaultOpenLevel` percentage attribute hit 100  then prior logic just added 1 to prior value which gave us 101, exceeding valid range
- **Fix**: Integer logic now wraps within valid range

**CameraAvStreamManagement** 
- **Root cause**: `Viewport` struct attribute unsupported by test's simple value generation
- **Fix**: Skip complex struct types as these are beyond the scope of IDM tests and covered more in depth by ACE tests.

### Specific Attributes now in the ATTRIBUTES_WITH_WRITE_CONSTRAINTS list:

**AccessControl (0x001F) ACL attribute**
- **Spec reference**: 9.10.4 - ACL Attribute
- **Reason**: Writing empty list to `Acl` attribute causes loss of all permissions, preventing test recovery (known from ACL test development), so this would need an additional check to be added and to pass this test, easiest to just skip this one by adding it to the exclusion list as the ACL tests thoroughly test the ACL attributes and functionaltiy.

**NetworkCommissioning (0x0031) InterfaceEnabled Attribute**
- **Spec reference**: 11.8.6.5 - InterfaceEnabled Attribute
- **Reason**: `InterfaceEnabled` write returns `Failure` (Status 1) per spec - "may be protected to prevent network disconnection". Added this to the exclusion list to keep from abnormal behavior from occurring if we turn this toggle this bool to false accidentally.

## Conclusion

The improved test now handles:
- Type-aware value generation with constraint validation
- Proper integer range wrapping
- String length constraints
- Struct type detection and skipping

 Only 2 specific attributes remain excluded in the ATTRIBUTES_WITH_WRITE_CONSTRAINTS 
 list, and for legitimate operational/design reasons documented in the spec, no longer noticed to be subscription support issues with the INVALID_ACTION error as mentioned previously.

j-ororke replied:
## Current Test Step 10 Validation

You'll notice currently that Test Step 10 is limited to 25 clusters and 20 attribute changes. The test step says "change all **or few** of the attributes", which allows for practical constraints which we might be reaching. The 25 cluster limit is due to device memory exhaustion during priming report generation (we currently appear to be hitting `CHIP Error 0x0000000B: No memory` above that), though this still exceeds the spec's minimum requirement of 3 as mentioned in spec (2.11.2.2. Subscribe Interaction Limits). The 20 attribute change limit keeps sequential write duration well under the set subscription's MaxInterval to avoid timeouts. We could use batch writes to increase this, but sequential writes make report verification much clearer if needed for debugging. Despite these limits, the test still validates wildcard subscriptions across 25 different clusters, multiple endpoints, most attribute types (string, int, bool, list, enum), and confirms every changed attribute sends a subscription report, which I hope satisfies the test plan's intent while respecting real-world device constraints. 
Please let me know if you want me to change this and I will be more than happy to do so!

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
No longer noticing memory issues after latest changes, updated test step 10 to remove constraint of only testing 25 clusters and 20 attribute writes

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Adding TC_IDM_4_3 and TC_IDM_2_2 test modules to slow tests with their local run durations

j-ororke committed:
restore asyncio sleeps for cleanup after subscriptions are shutdown

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

restyled-commits committed:
Restyled by ruff

j-ororke committed:
Resolves linting errors, resolves test step 4 timing issue, and resolves test step 10 latency issue

restyled-commits committed:
Restyled by ruff

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving further linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving ever more linting errors

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Resolving even more linting errors

restyled-commits committed:
Restyled by ruff

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Update TC_IDM_4_3.py

Updating imports due to framework changes

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1084 @@
+                            # Example: Location attribute of BasicInformation cluster default value is "XX", requires value to be max length of 2 uppercase letters.
+                            if len(cached_val) <= 2:
+                                # Location, country codes, short identifiers
+                                new_val = "US" if cached_val != "US" else "CA"
```
I'd actually suggest moving this to the list of untestable attributes - this is messing with the power tables and it's totally valid for the device to have unsupported regions.

j-ororke replied:
Thank you, I have moved the Location attribute of the Basic Information cluster to the untestable attributes list here now

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1084 @@
+                            # (e.g., InvalidValue, ConstraintError - as long as it's not UnsupportedAccess)
+                            log.info(f"{test_step}: Write to {attribute.__name__} returned {resp[0].Status}")
+
+                    except Exception as e:
```
Suggest removing this - you shouldn't be getting random exceptions when writing attributes even if you're getting errors returned, right?

cecille reviewed (commented):
This is looking fairly reasonable. I'd like to hold onto this one until 1.6 so we can verify against a bunch of different devices - I'm a bit worried that some of the writes will not work

Would you mind just locally building a bunch of examples and running this test as a first pass to make sure that the write / report tests work on what we can test of the standard clusters?

j-ororke committed:
Updating TC_IDM_4_3 in regards to comments from Cecille:
- Removed get_device_cluster var
- Moved Location attribute of Basic Information cluster to list of untestable attributes
- Removed try and except clause from the write attribute code block section

j-ororke committed:
Updating code to respond to suggestions from Cecille on this PR:
- Adding write attributes back to initial values except for NumberOfRinses attribute where we get an InvalidInState error
- Updating time.sleep() to asyncio.sleep() to resolve ruff check that was returning ASYNC251 errors.
- Updated to using event subscription in test step 1 to resolve ruff check ASYNC110 issue

restyled-commits committed:
Restyled by ruff

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Remove TC_IDM_4_3 yaml script and reference to it in manualTests.json

j-ororke committed:
Updated amount of clusters and attributes needing to be checked in test step 10 as some apps dont contain 40 clusters and 100 attributes, so have lowered required amount to being greater than 10 clusters and 50 attributes, this now passes on the following apps: all-clusters, lock, ota-requestor, bridge, microwave-oven, rvc, network-manager, energy-management

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

raul-marquez-csa reviewed (approved):
Great job on getting this one to the finish line.
LGTM

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Updated string-keyed dict to a dataclass, changed to calls for those to using the dataclass, validated no issues noticed after change

restyled-commits committed:
Restyled by isort

j-ororke commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1059 @@
+                            f"{test_step}: [{changed_count}] Changed {attribute.__name__} (0x{attribute_id:04X}) on endpoint {endpoint_id}, cluster 0x{cluster_id:04X}: {cached_val} -> {new_val}")
+
+                        # Track this change for verification
+                        changed_attributes.append({
```
Thank you, I have updated this here now!

j-ororke committed:
Removed if not check for NumberOfRinses as issue with that attribute is no longer being observed

j-ororke commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1059 @@
+                old_value = change['old_value']
+
+                # We have an issue with the NumberOfRinses attribute, where we continuously get InvalidInState error.
+                if attr.__name__ != "NumberOfRinses":
```
So I was able to remove the if not check for the NumberOfRinses attribute as the issue is no longer reproducing after updating to the latest code from master, this might have been fixed in [43171](https://github.com/project-chip/connectedhomeip/pull/43171).

j-ororke committed:
simplify MRP session param reads as per PR comment

j-ororke committed:
Updating default intervals from 0 to 500 for idle MRP and 300 for active MRP according to ReliableMessageProtocolConfig.h

j-ororke committed:
Adding comment for justification of idle and active MRP interval defaults being used to support why we used those values

j-ororke committed:
Replace wildcard report dict payloads with dataclasses

j-ororke committed:
Updating sleep for test step 3 sub report to a bounded queue wait which is more deterministic and event-driven

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        return device_clusters
+
+    # Test configuration
+    min_interval_floor_sec: int = 1
```
```suggestion
    min_interval_floor_sec: int = 0
```
recommendation for ICDs

j-ororke replied:
Thank you! Updated!

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+            for cluster_class, attributes in clusters.items():
+                cluster_id = cluster_class.id
+                # Get writable attributes for this cluster from endpoints_tlv data
+                if endpoint_id not in self.endpoints_tlv:
```
does this happen? It seems like if there's a mismatch, there's something strange about this device and this shoudl probably cause a test failure.

j-ororke replied:
Good point. Agreed, if priming_data references an endpoint/cluster missing from endpoints_tlv on the same DUT, that's a real inconsistency and shouldn't be silently swallowed. I'll turn both continues into asserts.assert_in(...) calls with messages that name the offending endpoint/cluster. 
Just a heads up this function was moved to idm_support.py to allow these functions to be shareable once merged, so the fix here will land there instead.

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+                if endpoint_id not in self.endpoints_tlv:
+                    continue
+
+                if cluster_id not in self.endpoints_tlv[endpoint_id]:
```
same

j-ororke replied:
Thank you, this change has been applied and verified no current issue noticed at this time, but if something comes up in the future then this will catch it, instead of silently swallowing it.

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+                    # Get current value from priming data
+                    cached_val = attributes[attribute]
+
+                    # Skip if value decode failed
```
Is this happening? It shouldn't be.

j-ororke replied:
Agreed, it shouldn't be happening. A ValueDecodeFailure in the subscription priming data means the DUT returned an undecodable value for an attribute it claimed to support in AttributeList, which is a DUT bug. Replaced the continue with an asserts.assert_false(...) that fails the test and reports the endpoint, cluster ID, and attribute name. This matches how TC_AccessChecker and TC_ACE_1_4 handle the same situation. 
Please note this landed in idm_support.py since the helper was moved there to make this shareable.

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+                    elif isinstance(cached_val, bool):
+                        # Boolean attribute - flip the value to trigger actual change
+                        new_val = not cached_val
+                    elif isinstance(cached_val, (int, float)):
```
Can you pull this from the data model constraints? This feels fragile.

j-ororke replied:
That'd be the better long-term direction, but I don't think it's available today through the spec parser. XmlAttribute doesn't surface attribute-level <constraint> data, and XmlDataTypeComponent.constraints is per-struct-field / per-enum-value rather than per-attribute.

For something like DefaultOpenLevel (uint8, valid range 0-100) we'd need to extend _parse_field_constraints (or a sibling) to also handle attribute elements and hang the result off XmlAttribute.

In the meantime, this heuristic just needs to produce some value the DUT will accept as a change. Writes that get rejected with ConstraintError / InvalidValue are logged and skipped, and the final verification only runs over changed_attributes, which we only append on Status.Success, so an out-of-range candidate produces a noisy log line, not a false failure.

Happy to file a follow-up issue to extend the spec parser and rework this on top of it. OK to keep that out of this PR?

cecille commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        log.info(f"Changed and verified {changed_count} attribute(s)")
+
+        handler_step10.shutdown()
+        await asyncio.sleep(1)  # Allow time for subscription teardown to complete
```
Do you know what's lagging here that causes this sleep to be required?

j-ororke replied:
The lag is in the Python binding's Shutdown() path. pychip_ReadClient_ShutdownSubscription ([src/controller/python/matter/clusters/attribute.cpp:546](https://github.com/project-chip/connectedhomeip/blob/a3150538087bd10a03092dc3b4f94594a40cddcf/src/controller/python/matter/clusters/Attribute.py#L546)) just dispatches InteractionModelEngine::ShutdownSubscription onto the Matter thread and returns - the return value is even explicitly ignored. There's no awaitable on the Python side that resolves once the ReadClient teardown actually completes, so when the next test step kicks off a new subscription right after Shutdown(), it can race with leftover in-flight attribute callbacks from the previous subscription. The wildcard steps (8/10/11) hit this most because of the sheer volume of pending callbacks. The sleep(1) is empirical padding to let the Matter thread drain.

A better fix would be to update the binding so ShutdownSubscription() exposes a future Python can await. I think that's worth tracking as a follow-up issue rather than blocking this PR, happy to file it or add it here if wanted, but if I we do it would expand this PR further.

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+    max_interval_ceiling_sec: int = 3
+    root_node_endpoint: int = 0
+
+    def get_mrp_retransmission_timeout_sec(self, dev_ctrl) -> float:
```
Is this `get_mrp_retransmission_timeout_sec` function and both `get_writable_attributes_for_cluster` and `change_writable_attributes_and_verify_reports` exclusive for IDM 4.3? Or are these also useful for other IDM tests? Or tests outside of IDM?

Just wondering if these can be moved to a supporting class.

j-ororke replied:
Good idea, I can move these into idm_support.py.

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+
+        attr_handler_step1.cancel()
+
+        # New Step 2: Basic attribute change and report timing
```
```suggestion
        # Step 2: Basic attribute change and report timing
```
"New" comment probably left during development.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        attr_handler_step1.cancel()
+
+        # New Step 2: Basic attribute change and report timing
+        # (This was originally test step 2 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        attr_handler_step2.cancel()
+
+        # Step 3: Multiple subscriptions with KeepSubscriptions=False
+        # (This was originally test step 11 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        attr_handler_step3_second.cancel()
+
+        # Step 4: MinInterval/MaxInterval timing
+        # (This was originally test steps 12 and 13 in the test plan, just appears to have been combined into one test step, which was a wise design decision)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        attr_handler_step4.cancel()
+
+        # Step 5: KeepSubscriptions=True preserves first subscription
+        # (This was originally test step 14 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        attr_handler_step5_second.cancel()
+
+        # Step 6: KeepSubscriptions=False cancels first subscription
+        # (This was originally test step 15 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        attr_handler_step6_second.cancel()
+
+        # Step 7: Subscription to attribute and events
+        # (This was originally test step 16 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        attr_handler_step7.cancel()
+
+        # Step 8: Attribute wildcard subscription
+        # (This was originally test step 17 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        await asyncio.sleep(1)  # Allow time for subscription teardown to complete
+
+        # Step 9: Attribute on cluster from all endpoints
+        # (This was originally test step 18 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        await asyncio.sleep(1)  # Allow time for subscription teardown to complete
+
+        # Step 10: All attributes from all clusters from all endpoints
+        # (This was originally test step 19 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        await asyncio.sleep(1)  # Allow time for subscription teardown to complete
+
+        # Step 11: All attributes from all clusters on an endpoint
+        # (This was originally test step 20 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

raul-marquez-csa commented on src/python_testing/TC_IDM_4_3.py:
```diff
@@ -0,0 +1,1062 @@
+        await asyncio.sleep(1)  # Allow time for subscription teardown to complete
+
+        # Step 12: All attributes from specific cluster on all endpoints
+        # (This was originally test step 21 in the test plan)
```
```suggestion
```
Comment probably left during development, suggest removing.

j-ororke replied:
Thank you, removed!

j-ororke committed:
Apply suggestions from code review by Raul

Removes unnecessary comments from the test module, thank you!

Co-authored-by: Raul Marquez <130402456+raul-marquez-csa@users.noreply.github.com>

raul-marquez-csa commented on src/python_testing/matter_testing_infrastructure/matter/testing/event_attribute_reporting.py:
```diff
@@ -687,3 +703,165 @@ def await_duration_asserting_no_forbidden(
+        """
+        report_key = (endpoint_id, cluster_type, attribute_type)
+        with self._lock:
+            return report_key in self._attribute_reports and len(self._attribute_reports[report_key]) > 0
```
The second condition `len(self._attribute_reports[report_key]) > 0` can never be False if the first condition `report_key in self._attribute_reports` is True.

This is due to:
```
        with self._lock:
            if report_key not in self._attribute_reports:
                self._attribute_reports[report_key] = []
            self._attribute_reports[report_key].append(WildcardAttributeReportSnapshot(
```
A key existing in the dict implies at least one element in its list, in the code block both paths add a key.

And after a `reset()` or `_attribute_reports.clear()`, it removes all the keys, which would make `report_key in self._attribute_reports` resolve to False and `report_key in self._attribute_reports` is never reached.

So:
```return report_key in self._attribute_reports and len(self._attribute_reports[report_key]) > 0```

Can be simplified to:
```return report_key in self._attribute_reports```

j-ororke replied:
Good catch, you're right. 
Both code paths that touch self._attribute_reports[report_key] add the key and append in the same locked block, so an empty list under an existing key isn't reachable. 
Simplified this here now to return report_key in self._attribute_reports. 
Thank you!

j-ororke committed:
Adding suggestion from Cecille

Set default value to 0 for min_int_floor_sec var.

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
Address PR feedback from Cecille and Raul:
- Move IDM 4.3 helpers (MRP timeout, writable attribute discovery,
  change-and-verify) and ChangedAttribute dataclass into IDMBaseTest in
  support_modules/idm_support.py so other IDM tests can reuse them.
- Have TC_IDM_4_3 inherit from IDMBaseTest, BasicCompositionTests to
  match the existing IDM test pattern.
- Fail loudly when subscription priming data references endpoints or
  clusters missing from the wildcard composition read instead of
  silently skipping.
- Fail loudly on ValueDecodeFailure in priming data; the DUT claimed
  the attribute is supported, so an undecodable value is a DUT bug.
- Drop redundant len(...) > 0 check in
  WildcardAttributeSubscriptionHandler.was_attribute_reported.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

cecille commented on src/python_testing/test_metadata.yaml:
```diff
@@ -193,6 +193,10 @@ slow_tests:
     - { name: TC_FAN_3_1.py, duration: 15 seconds }
+    - { name: TC_IDM_2_2.py, duration: 3 minutes }
+    - { name: TC_IDM_4_3.py, duration: 4 minutes }
+    - { name: TC_JFPKI_2_4.py, duration: 5 minutes }
```
were these added intentionally or is this a weird merge?

j-ororke replied:
Thank you for catching this! 
This is weird, it must have happened during merge somehow. I can speak for the IDM tests as we wrote both of those, and they should be in here as they do take around that amount of time to run normally.

However, I do not know about the JFPKI test listed here as it appears to have been renamed or removed since it doesn't exist in master currently, so removing that one from this list. The JFADMIN tests appear to have been moved to the bottom of this list so removing adding it twice, since the location of it has changed.

j-ororke committed:
Removing duplicate JFADMIN test entry and removing non-existent JFPKI test from slow_tests list in test_metadata.py

j-ororke committed:
Merge branch 'master' into update_TC_IDM_4_3_test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
